doc/neps/nep-0035-array-creation-dispatch-with-array-function.rst

```diff
@@ -8,16 +8,176 @@ NEP 35 — Array Creation Dispatching With __array_function__
 :Status: Draft
 :Type: Standards Track
 :Created: 2019-10-15
-:Updated: 2020-08-06
+:Updated: 2020-08-17
 :Resolution:
 
 Abstract
 --------
 
 We propose the introduction of a new keyword argument ``like=`` to all array
-creation functions to permit dispatching of such functions by the
-``__array_function__`` protocol, addressing one of the protocol shortcomings,
-as described by NEP-18 [1]_.
+creation functions, this argument permits the creation of an array based on
+a non-NumPy reference array passed via that argument, resulting in an array
+defined by the downstream library implementing that type, which also implements
+the ``__array_function__`` protocol. With this we address one of that
+protocol's shortcomings, as described by NEP 18 [1]_.
+
+Motivation and Scope
+--------------------
+
+Many are the libraries implementing the NumPy API, such as Dask for graph
+computing, CuPy for GPGPU computing, xarray for N-D labeled arrays, etc. All
+the libraries mentioned have yet another thing in common: they have also adopted
+the ``__array_function__`` protocol; a protocol that allows NumPy to understand
+and treat downstream objects as if they are the native ``numpy.ndarray`` object.
+Hence the community while using various libraries still benefits from a unified
+NumPy API. This not only brings great convenience for standardization but also
+removes the burden of learning a new API and rewriting code for every new
+object. In more technical terms, this mechanism of the protocol is called a
+"dispatcher", which is the terminology we use from here onwards when referring
+to that.
+
+
+.. code:: python
+
+    x = dask.array.arange(5)    # Creates dask.array
+    np.diff(x)                  # Returns dask.array
+
+Note above how we called Dask's implementation of ``sum`` via the NumPy
+namespace by calling ``np.diff``, and the same would apply if we had a CuPy
+array or any other array from a library that adopts ``__array_function__``.
+This allows writing code that is agnostic to the implementation library, thus
+users can write their code once and still be able to use different array
+implementations according to their needs.
+
+Obviously, having a protocol in-place is useful if the arrays are created
+elsewhere and let NumPy handle them. But still these arrays have to be started
+in their native library and brought back. Instead if it was possible to create
+these objects through NumPy API then there would be an almost complete
+experience, all using NumPy syntax. For example, say we have some CuPy array
+``cp_arr`` , and want a similar CuPy array with identity matrix. We could still
+write the following:
+
+.. code:: python
+     x = cupy.identity(3)
+
+Instead, the better way would be using to only use the NumPy API, this could now
+be achieved with:
+
+.. code:: python
+    x = np.identity(3, like=cp_arr)
```

Comment:
Perhaps out of scope, but... Should this copy the dtype from the `like` argument? How about cupy-specific metadata?

Reply:
This is a valid comment. It should create the array on the same GPU as the original. In the context of Dask, it would copy the chunk-sizes, and for Sparse, it would copy the format.

However, I don't think this detail needs to be discussed, and currently `__array_function__` only passes through the type. One can cheat and copy from `self` though, as it isn't a `classmethod`.

Reply:
I think the "same GPU" remark is most salient here. I think we should clarify the wording of "does no processing or copying" to either allow for this, or deliberately declare it out of scope.

Reply:
> Perhaps out of scope, but... Should this copy the dtype from the like argument? How about cupy-specific metadata?

Array creation functions will generally specify this type of information via arguments, whereas `like=` is used to identify where to dispatch. I don't think we would be able to detect that, and I'm even less sure we would want that, as where we would expect `like=` to be used to already have the correct arguments defined by the library using it.

> This is a valid comment. It should create the array on the same GPU as the original.
> ...
> I think the "same GPU" remark is most salient here

This is not possible, GPU arrays don't store this information, those arrays only know the pointer. The GPU where it's created is based on the current context selected by the library, such as `with cupy.cuda.Device(0):` . With that said, it's _always_ the responsibility of the code to define that, regardless of what this NEP is doing.